### DISCUSSION TITLE
OpenSCAD to JSCAD

### DISCUSSION THREAD

**Discussion by napter** [jscad/OpenJSCAD.org#1291] [General]
Hi all!

I've been working on an openscad to jscad converter since the official one has been abandoned. It works on a subset of the openscad grammar but can be expanded further. Would you be interested in using or contributing this project?

I also want to show off our CAD app where this will be used: CADit.app. You can run jscad code in there and then manipulate it visually. We are considering a reverse option, from visual to jscad. We will also be adding a library where people can share JSCad designs. 

Let me know what you think!
-Nathan

**Comment by platypii** [comment]
Very cool! I would love to have this functionality in JSCAD working to convert from openscad. I've wanted to try to fix the existing `scad-deserializer` for a while (not enough time in the day).

I think it would make JSCAD much more accessible to a whole world of people who have used openscad, have existing designs in scad, and are maybe interested in JSCAD, but re-writing everything in one go might be too much to ask.

Are you thinking this might replace the broken `scad-deserializer` package? I'm sure we would be happy to consider a PR to merge it back in if that's something you'd be interested in

**Reply by napter** [reply]
I am not familiar with that package, but yes this is to translate code from openscad to jscad. It currently works, but on a subset of the language. It mostly needs brute manual labor to do all the mappings.

**Comment by hrgdavor** [comment]
this could be useful https://github.com/alufers/openscad-parser

**Comment by pca006132** [comment]
Converter and interpreter are different. For converters, you may need some sort of type inference for you to find out the type of the expression, which can be non-trivial. For interpreter, you just execute the script, which is a lot easier.

The main problem for interpreter is that the output may not be very readable, e.g. turns everything into a sequence of function calls without much structure.

**Reply by hrgdavor** [reply]
converting openscad to jscad would be very interesting if one could convert openscad code to JS equivalent with api calls to jscad. That would allow jscad to tap in into many nice libs from openscad and benefit from their lively community.
if not able to perfectly convert, at some level of success it becomes a nifty tool for porting code. I could theoretically help, but my TODO list for other jscad stuff is soo large, I can only spare some time to track progress of such effort, and comment a bit.

**Comment by napter** [comment]
I've gotten a working POC and have published it to https://github.com/cookiecad/Openscad-to-JSCAD-converter/

The general idea is to use an openscad grammar to parse the openscad code into a syntax tree, then use the tree to jenerate jscad code. In order to make the resulting code more readable or when jscad is missing an equivalent operation the code adds helper functions.

I used the openscad version of the honeycomb storage wall (on printables) as my test case and it successfully converts it to jscad. Each language feature is added separately so there are still many more to add.

If anyone wants to try it out or contribute let me know. I haven't had time to work on it recently but would love to make more progress. Index.js is the starting point.

As a side note - you can run jscad code with auto-completion and typechecking in our site, cadit.app - let me know what you think!

**Comment by z3dev** [comment]
@napter I just published a bridge API from OpenSCAD to JSCAD. It's still in progress but functional. There are some very string things in OpenSCAD, so maybe this will help with the 'helper' functions, etc. I'd be glad to help contribute to your project as well.

https://github.com/jscad-community/jscad-scad-api

My goal is to take all OpenSCAD projects on Thingiverse and convert to JSCAD. 😃  HA HA HA!

**Reply by napter** [reply]
That is my goal too! Jscad needs a larger library of projects.

**Comment by z3dev** [comment]
@napter I just finished a round of testing for jscad-scad-api. I spent time to verify that the functionality for the various functions follows OpenSCAD 😷 But anyway, there were a significant amount of fixes, and improvements.

How do you want to proceed? I think that the jscad-scad-api can be a good reference for the conversions. I can jump on to the converter now, and make changes. I think we can probably make some nice progress by combining effort.

**Reply by napter** [reply]
Yes, that sounds great!

**Comment by z3dev** [comment]
FYI, Thingiverse provides an API to access the list of designs, modify designs, and create designs.

https://www.thingiverse.com/developers
